lib/kinesis_client/stream/shard/producer.ex

```diff
@@ -131,49 +135,29 @@ defmodule KinesisClient.Stream.Shard.Producer do
   end
 
   @impl GenStage
-  def handle_info({:ack, _ref, [], failed_msgs}, state) do
-    Logger.debug("Retrying #{length(failed_msgs)} failed messages")
-
-    state =
-      case state.shard_closed_timer do
-        nil ->
-          state
-
-        timer ->
-          Process.cancel_timer(timer)
-          %{state | shard_closed_timer: nil}
-      end
-
-    {:noreply, failed_msgs, state}
+  def handle_info(msg, state) do
+    Logger.debug("ShardConsumer.Producer got an unhandled message #{inspect(msg)}")
+    {:noreply, [], state}
   end
 
-  @impl GenStage
-  def handle_info({:ack, _ref, successful_msgs, failed_msgs}, state) do
-    %{metadata: %{"SequenceNumber" => checkpoint}} = successful_msgs |> Enum.reverse() |> hd()
-
-    :ok = update_checkpoint(state, checkpoint)
+  def get_top_checkpoint(successful_msgs, failed_msgs) do
+    success_checkpoint = get_top_checkpoint(successful_msgs)
+    failed_checkpoint = get_top_checkpoint(failed_msgs)
+    checkpoint = Enum.sort([success_checkpoint, failed_checkpoint], :desc) |> hd()
 
-    Logger.debug(
-      "Acknowledged #{length(successful_msgs)} messages, Retrying #{length(failed_msgs)} failed messages"
-    )
+    if checkpoint == "-1" do
+      raise "Error getting highest checkpoint from acknowledged messages: #{inspect(successful_msgs)} #{inspect(failed_msgs)}"
+    end
 
-    state =
-      case state.shard_closed_timer do
-        nil ->
-          state
+    checkpoint
+  end
 
-        timer ->
-          Process.cancel_timer(timer)
-          %{state | shard_closed_timer: nil}
-      end
+  def get_top_checkpoint([]), do: "0"
 
-    {:noreply, failed_msgs, state}
-  end
+  def get_top_checkpoint(msgs) do
+    %{metadata: %{"SequenceNumber" => checkpoint}} = msgs |> Enum.reverse() |> hd()
```

Comment:
While we're in here, I wonder if we want to make this more resilient? E.g. guard against but map shape, empty message list etc.